### ISSUE TITLE
removing an object using dbstore implementation actually triggers a seg-fault in radosgw

### ISSUE THREAD

**Issue by giubacc** [aquarist-labs/s3gw-core#7]
Started rgw with:  `radosgw --rgw-backend-store=dbstore -i foo --debug-rgw 15 --no-mon-config --rgw-data ${workspaceFolder}/rgw --run-dir ${workspaceFolder}/rgw -d`

- created bucket with: `s3cmd --no-ssl -c s3cmd.cfg mb s3://test` 
- put a file into that bucket: `s3cmd --no-ssl -c s3cmd.cfg put myfile s3://prova`
- tried to remote that file with: `s3cmd --no-ssl -c s3cmd.cfg rm s3://prova/myfile`

`rm` command actually causes `radosgw` to seg-fault.

call stack:

```
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::capacity(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > * const this) (/usr/include/c++/9/bits/basic_string.h:997)
std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > * const this, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > & __str) (/usr/include/c++/9/bits/basic_string.tcc:260)
libradosgw.so.2!std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::assign(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > & __str, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > * const this) (/usr/include/c++/9/bits/basic_string.h:1368)
libradosgw.so.2!std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator=(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > & __str, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > * const this) (/usr/include/c++/9/bits/basic_string.h:700)
libradosgw.so.2!rgw_bucket::operator=(rgw_bucket * const this) (/workspaces/ceph/src/rgw/rgw_basic_types.h:289)
libradosgw.so.2!rgw_obj::operator=(rgw_obj * const this) (/workspaces/ceph/src/rgw/rgw_common.h:1799)
libradosgw.so.2!RGWObjState::operator=(RGWObjState * const this) (/workspaces/ceph/src/rgw/rgw_sal.h:104)
libradosgw.so.2!rgw::store::DB::Object::get_obj_state(rgw::store::DB::Object * const this, const DoutPrefixProvider * dpp, const RGWBucketInfo & bucket_info, const rgw_obj & obj, bool follow_olh, RGWObjState ** state) (/workspaces/ceph/src/rgw/store/dbstore/common/dbstore.cc:1229)
libradosgw.so.2!rgw::sal::DBObject::get_obj_state(rgw::sal::DBObject * const this, const DoutPrefixProvider * dpp, RGWObjState ** pstate, optional_yield y, bool follow_olh) (/workspaces/ceph/src/rgw/rgw_sal.h:1094)
libradosgw.so.2!RGWDeleteObj::execute(RGWDeleteObj * const this, optional_yield y) (/workspaces/ceph/src/rgw/rgw_op.cc:4932)
...
```

A possible explanation of this crash:

rgw_op.cc

```
void RGWDeleteObj::execute(optional_yield y)
{
  if (!s->bucket_exists) {
    op_ret = -ERR_NO_SUCH_BUCKET;
    return;
  }

  if (!rgw::sal::Object::empty(s->object.get())) {
    uint64_t obj_size = 0;
    std::string etag;
    {
      RGWObjState* astate = nullptr;
      bool check_obj_lock = s->object->have_instance() && s->bucket->get_info().obj_lock_enabled();

      op_ret = s->object->get_obj_state(this, &astate, s->yield, true);
```

dbstore.cc

```
int DB::Object::get_obj_state(const DoutPrefixProvider *dpp,
                              const RGWBucketInfo& bucket_info, const rgw_obj& obj,
                              bool follow_olh, RGWObjState **state)
{
  int ret = 0;

  DBOpParams params = {};
  RGWObjState* s;
  store->InitializeParams(dpp, &params);
  InitializeParamsfromObject(dpp, &params);

  ret = store->ProcessOp(dpp, "GetObject", &params);

  if (ret) {
    ldpp_dout(dpp, 0)<<"In GetObject failed err:(" <<ret<<")" << dendl;
    goto out;
  }

  if (!params.op.obj.state.exists) {
    return -ENOENT;
  }

  s = &params.op.obj.state;
  /* XXX: For now use state->shadow_obj to store ObjectID string */
  s->shadow_obj = params.op.obj.obj_id;

  **state = *s;
```

the actual implementation of `get_obj_state` for `dbstore` never allocates `*state`.
this causes  `**state = *s` to trigger a sig-fault.

**Comment by jecluis** [comment]
I confirm this is happening

**Comment by jecluis** [comment]
This also happens when we remove a bucket recursively with `s3cmd rb --recursive s3://foo`.

**Comment by jecluis** [comment]
```
➜ joao@algazarra rgw.foo git:(s3gw) s3cmd -c ./s3.cfg rm s3://foo/s3cfg2
delete: 's3://foo/s3cfg2'
➜ joao@algazarra rgw.foo git:(s3gw) s3cmd -c ./s3.cfg put s3.cfg s3://foo/s3cfg3
upload: 's3.cfg' -> 's3://foo/s3cfg3'  [1 of 1]
 2281 of 2281   100% in    0s    50.41 KB/s  done
➜ joao@algazarra rgw.foo git:(s3gw) s3cmd -c ./s3.cfg rm s3://foo/s3cfg3        
^CSee ya!
```

**Comment by giubacc** [comment]
Uh, this is quite strange! Would be worth maybe to add a print just before `**state = *s` to verify that `*state` is actually always `nullptr`. From a static check it would seems so.

**Comment by giubacc** [comment]
I checked the implementation for rados backend.

rgw_rados.cc

```
int RGWRados::get_obj_state_impl(const DoutPrefixProvider *dpp, RGWObjectCtx *rctx,
				 const RGWBucketInfo& bucket_info, rgw::sal::Object* obj,
                                 RGWObjState **state, RGWObjManifest** manifest,
				 bool follow_olh, optional_yield y, bool assume_noent)
{
  if (obj->empty()) {
    return -EINVAL;
  }

  bool need_follow_olh = follow_olh && obj->get_obj().key.instance.empty();
  *manifest = nullptr;

  RGWObjStateManifest *sm = rctx->get_state(obj->get_obj());
  RGWObjState *s = &(sm->state);
  ldpp_dout(dpp, 20) << "get_obj_state: rctx=" << (void *)rctx << " obj=" << obj << " state=" << (void *)s << " s->prefetch_data=" << s->prefetch_data << dendl;
  *state = s;
```
The intention here is to make `s` available outside `get_obj_state` function.
We can see that the original intention seems not to allocate `*state` with a new object and then copy on it `*s`.
Otherwise, In the dbstore implementation, `**state = *s;` seems to indicate that the code expects `*state` to be already allocated by the `get_obj_state` caller, in fact, it claims to actually copy `*s` on it.
I would say that changing `**state = *s` with `*state = s;` could be a fix.

**Comment by giubacc** [comment]
Well, my last comment was quite optimistic, the fix I proposed simply cannot work: the issue here is actually far more deep. 
Rados backend implementation relies on a concept of "object context", that basically is a map containing some info regarding a certain persisted-object.
Those infos include also the "state" of an object; and in fact, that state is retrieved by `src/rgw/rgw_sal_rados.cc` querying that map.
Because of the design of the SAL interface, that "state" has to _necessary_ be an object living in the `heap` and not on the `stack`.
The current implementation in the dbstore is flawed because the "state" object is living on the `stack` and allocated in a innermost function used by the SAL interface. 
So, there is currently no way to make that memory survive after the innermost function returns.
Allocating it in the `heap` is not possible either, because the abstract code in the interface does not contemplate its deallocation.
We would have to implement an analogous object context like in the rados backend implementation.
Said this, and because this is clearly a macro-issue, I would say that probably the dbstore codebase is in an early stage of its development and ultimately far from be mature.

**Comment by giubacc** [comment]
After have discussed with the team what could be the best approach for the MVP scope; we agreed on reporting this issue upstream.
We will then try to provide a patch with the aim to solely make DELETE operation working for the MVP.

**Comment by jecluis** [comment]
For a nasty fix to get around the problem quickly, can't we simply allocate the required state in the heap and ignore the actual cleanup? Or use a shared_ptr?

**Comment by giubacc** [comment]
I try simply allocating it and it worked.
I'm trying to change the signature of SAL interface, basically adding as last parameter a reference for a local boolean that will hint the caller if a delete must be performed.
Should we then PR this change against our [`s3gw` branch](https://github.com/aquarist-labs/ceph/tree/s3gw) ?

**Comment by jecluis** [comment]
Fixed by aquarist-labs/ceph#1.

Subsequent issues should have their own issues filed.
